primitives/trie/src/trie_codec.rs

```diff
@@ -0,0 +1,192 @@
+use sp_std::vec::Vec;
+use trie_db::Trie;
+
+type VerifyError<L> = crate::VerifyError<TrieHash<L>, Box<TrieError<L>>>;
+
+fn verify_error<L: TrieConfiguration>(error: Box<TrieError<L>>) -> VerifyError<L> {
+	VerifyError::<L>::DecodeError(error)
+}
+
+/// Decode a compact proof.
+///
+/// Takes as input a destination `db` for decoded node and `encoded`
+/// an iterator of compact encoded nodes.
+///
+/// Child trie are decoded in order of child trie root present
+/// in the top trie.
+pub fn decode_compact<'a, L, DB, I>(
+	db: &mut DB,
+	encoded: I,
+	expected_root: Option<&TrieHash<L>>,
+) -> Result<TrieHash<L>, VerifyError<L>>
+	where
+		L: TrieConfiguration,
+		DB: HashDBT<L::Hash, trie_db::DBValue> + hash_db::HashDBRef<L::Hash, trie_db::DBValue>,
+		I: IntoIterator<Item = &'a [u8]>,
+{
+	let mut nodes_iter = encoded.into_iter();
+	let (top_root, _nb_used) = trie_db::decode_compact_from_iter::<L, _, _, _>(
+		db,
+		&mut nodes_iter,
+	).map_err(verify_error::<L>)?;
+
+	if let Some(expected_root) = expected_root {
+		if expected_root != &top_root {
+			return Err(VerifyError::<L>::RootMismatch(expected_root.clone()));
+		}
+	}
+
+	let mut child_tries = Vec::new();
+	{
+		// fetch child trie roots
+		let trie = crate::TrieDB::<L>::new(db, &top_root).map_err(verify_error::<L>)?;
+
+		let mut iter = trie.iter().map_err(verify_error::<L>)?;
+
+		let childtrie_roots = sp_core::storage::well_known_keys::DEFAULT_CHILD_STORAGE_KEY_PREFIX;
+		if iter.seek(childtrie_roots).is_ok() {
+			loop {
+				match iter.next() {
+					Some(Ok((key, value))) if key.starts_with(childtrie_roots) => {
+						// we expect all default child trie root to be correctly encoded.
+						// see other child trie functions.
+						let mut root = TrieHash::<L>::default();
+						// still in a proof so prevent panic
+						if root.as_mut().len() != value.as_slice().len() {
+							return Err(VerifyError::<L>::RootMismatch(Default::default()));
+						}
+						root.as_mut().copy_from_slice(value.as_ref());
+						child_tries.push(root);
+					},
+					// allow incomplete database error: we only
+					// require access to data in the proof.
+					Some(Err(error)) => match *error {
+						trie_db::TrieError::IncompleteDatabase(..) => (),
+						e => return Err(VerifyError::<L>::DecodeError(Box::new(e))),
+					},
+					_ => break,
+				}
+			}
+		}
+	}
+
+	if !HashDBT::<L::Hash, _>::contains(db, &top_root, EMPTY_PREFIX) {
+		return Err(VerifyError::<L>::IncompleteProof);
+	}
+
+	let mut previous_extracted_child_trie = None;
+	for child_root in child_tries.into_iter() {
+		if previous_extracted_child_trie == None {
+			let (top_root, _) = trie_db::decode_compact_from_iter::<L, _, _, _>(
+				db,
+				&mut nodes_iter,
+			).map_err(verify_error::<L>)?;
+			previous_extracted_child_trie = Some(top_root);
+		}
+
+		// we allow skipping child root by only
+		// decoding next on match.
+		if Some(child_root) == previous_extracted_child_trie {
+			previous_extracted_child_trie = None;
+		}
```

Comment:
I'm confused about this logic. What is that doing?

Reply:
Same thing, it allows to have child root in the top proof without having a child proof to it.
In practice it will not really occurs (our child trie location are obtained from hash of similar size so embeding an additional content in proof will not happen), but in tests (possibly key containing other key) it could happen.
If for some reason we did not have child proof but it is expected, then it will fail during verification.
This is here to avoid user adding a child trie proof that without its root in the parent proof (in practice this will also fail during verification, but we see it earlier).